lib/routine/routine_test.go

```diff
@@ -99,3 +99,47 @@ func TestManager_StopAll(t *testing.T) {
 		require.False(r, mgr.IsRunning("run2"))
 	})
 }
+
+// Test IsRunning when routine is a blocking call that does not
+// immediately return when cancelled
+func TestManager_StopBlocking(t *testing.T) {
+	t.Parallel()
+	var runs uint32
+	var running uint32
+	unblock := make(chan struct{}) // To simulate a blocking call
+	mgr := NewManager(testutil.Logger(t))
+
+	// A routine that will be still running for a while after cancelled
+	run := func(ctx context.Context) error {
+		atomic.StoreUint32(&running, 1)
+		defer atomic.StoreUint32(&running, 0)
+		atomic.AddUint32(&runs, 1)
+		<-ctx.Done()
+		<-unblock
+		return nil
+	}
+
+	require.NoError(t, mgr.Start(context.Background(), "blocking", run))
+	retry.Run(t, func(r *retry.R) {
+		require.True(r, mgr.IsRunning("blocking"))
+		require.Equal(r, uint32(1), atomic.LoadUint32(&runs))
+		require.Equal(r, uint32(1), atomic.LoadUint32(&running))
+	})
+
+	doneCh := mgr.Stop("blocking")
+
+	// IsRunning should return false, however &running is still 1
+	retry.Run(t, func(r *retry.R) {
+		require.False(r, mgr.IsRunning("blocking"))
+		require.Equal(r, uint32(1), atomic.LoadUint32(&running))
+	})
+
```

Comment:
It would be really great to add a restart of the routine here

```suggestion
	require.NoError(t, mgr.Start(context.Background(), "blocking", func(ctx context.Context) { <-ctx.Done() }))
	defer mgr.Stop("blocking")

	retry.Run(t, func(r *retry.R) {
 		require.True(r, mgr.IsRunning("blocking"))
 		require.Equal(r, uint32(1), atomic.LoadUint32(&running))
 	})
```

I think this helps to ensure that restarting a cancelled routine allows replacement of the original one even if it is not actually finished running.

Note that in addition to this suggestion you would also need to change the IsRunning check in the future block as the routine would actually be running.